src/java/org/apache/cassandra/config/Converters.java

```diff
@@ -36,64 +36,77 @@
      * This converter is used when we change the name of a cassandra.yaml configuration parameter but we want to be
      * able to still use the old name too. No units involved.
      */
-    IDENTITY(null, o -> o, o-> o),
-    MILLIS_DURATION(Long.class,
-                    o -> SmallestDurationMilliseconds.inMilliseconds((Long) o),
-                    o -> ((SmallestDurationMilliseconds)o).toMilliseconds()),
-    MILLIS_DOUBLE_DURATION(Double.class,
-                           o ->  SmallestDurationMilliseconds.inDoubleMilliseconds((Double) o),
-                           o -> ((SmallestDurationMilliseconds)o).toMilliseconds()),
+    IDENTITY(null, null, o -> o, o-> o),
+    MILLIS_DURATION(Long.class, SmallestDurationMilliseconds.class,
+                    SmallestDurationMilliseconds::inMilliseconds,
+                    o -> o.toMilliseconds()),
+    MILLIS_DOUBLE_DURATION(Double.class, SmallestDurationMilliseconds.class,
+                           o -> Double.isNaN(o) ? SmallestDurationMilliseconds.inMilliseconds(0) : SmallestDurationMilliseconds.inDoubleMilliseconds(o),
+                           o -> (double) o.toMilliseconds()),
     /**
      * This converter is used to support backward compatibility for parameters where in the past -1 was used as a value
-     * Example: credentials_update_interval_in_ms = -1 and credentials_update_interval = null (quantity of 0ms) are equal.
+     * Example: credentials_update_interval_in_ms = -1 and credentials_update_interval = null are equal.
      */
-    MILLIS_CUSTOM_DURATION(Long.class,
-                           o -> (Long)o == -1 ? new SmallestDurationMilliseconds("0ms") : SmallestDurationMilliseconds.inMilliseconds((Long) o),
-                           o -> ((SmallestDurationMilliseconds)o).toMilliseconds() == 0 ? -1 : ((SmallestDurationMilliseconds)o).toMilliseconds()),
-    SECONDS_DURATION(Long.class,
-                     o -> SmallestDurationSeconds.inSeconds((Long) o),
-                     o -> ((SmallestDurationSeconds)o).toSeconds()),
+    MILLIS_CUSTOM_DURATION(Long.class, SmallestDurationMilliseconds.class,
+                           o -> o == -1 ? null : SmallestDurationMilliseconds.inMilliseconds(o),
+                           o -> o == null ? -1 : o.toMilliseconds()),
+    SECONDS_DURATION(Long.class, SmallestDurationSeconds.class,
+                     SmallestDurationSeconds::inSeconds,
+                     DurationSpec::toSeconds),
+    NEGATIVE_SECONDS_DURATION(Long.class, SmallestDurationSeconds.class,
+                              o -> o < 0 ? SmallestDurationSeconds.inSeconds(0) : SmallestDurationSeconds.inSeconds(o),
+                              DurationSpec::toSeconds),
```

Comment:
should be `Integer` and not `Long`

```
NEGATIVE_SECONDS_DURATION(Integer.class, SmallestDurationSeconds.class,
                              o -> o < 0 ? SmallestDurationSeconds.inSeconds(0) : SmallestDurationSeconds.inSeconds(o),
                              DurationSpec::toSecondsAsInt),
```

Old type was `int`

Reply:
Please check my other comment.